Allow `MapieRegressor` to use group split strategy

pidefrem opened the pull request:
# Description

Continuing the work done in [PR](https://github.com/scikit-learn-contrib/MAPIE/pull/386)
Allow the use of `GroupKFold` split (see https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.GroupKFold.html) and also the use of custom splits based on `StratifiedKFold` for example (see https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.StratifiedKFold.html#sklearn.model_selection.StratifiedKFold)

Fixes #202

## Type of change

- New feature (non-breaking change which adds functionality)

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] test_regression.test_results_with_constant_groups

# Checklist

- [X] I have read the [contributing guidelines](https://github.com/simai-ml/MAPIE/blob/master/CONTRIBUTING.rst)
- [X] I have updated the [HISTORY.rst](https://github.com/simai-ml/MAPIE/blob/master/HISTORY.rst) and [AUTHORS.rst](https://github.com/simai-ml/MAPIE/blob/master/AUTHORS.rst) files
- [X] Linting passes successfully : `make lint`
- [X] Typing passes successfully : `make type-check`
- [X] Unit tests pass successfully : `make tests`
- [X] Coverage is 100% : `make coverage`
- [X] Documentation builds successfully : `make doc`

thibaultcordier commented:
Hello @pidefre! Thank you for this contribution. I'll take the time to review it. Just to be in line with your proposal, I understand that in `Allow MapieRegressor to use group split strategy`, the `split` strategy refers to the way of doing cross-validation but that it is a cross conformal method.

pidefrem commented:
Hello @thibaultcordier, yes it refers to the split methods of the cross validator used during the fit of the MAPIE estimators. Please feel free to suggest any other description that you think is more suitable.

thibaultcordier commented on Makefile:
```diff
@@ -1,7 +1,7 @@
 
-lint:	
-	flake8 . --exclude=doc
+lint:
+	flake8 . --exclude=doc,.venv
```
Have you encountered a problem that justifies this addition to the file?

pidefrem replied:
Hello @thibaultcordier. By default the `lint` command in your Makefile will scan every folder in the current workspace. `.venv` is the default location when using vscode and virtual envs, so your `lint` command was also scanning every dependency in the virtual env folder.

LacombeLouis replied:
Hey @pidefrem,
Could you please remove the environment? You can also save the env somewhere else than in the MAPIE folder. 
Thank you!

thibaultcordier commented on mapie/estimator/estimator.py:
```diff
@@ -322,7 +322,11 @@ def _pred_multi(self, X: ArrayLike) -> NDArray:
         return y_pred_multi
 
-    def predict_calib(self, X: ArrayLike) -> NDArray:
+    def predict_calib(
+            self,
+            X: ArrayLike,
+            y: Optional[ArrayLike] = None,
+            groups: Optional[ArrayLike] = None) -> NDArray:
```
```suggestion
    def predict_calib(
        self,
        X: ArrayLike,
        y: Optional[ArrayLike] = None,
        groups: Optional[ArrayLike] = None
    ) -> NDArray:
```

thibaultcordier commented on mapie/subsample.py:
```diff
@@ -89,7 +89,8 @@ def split(
             yield train_index, test_index
 
-    def get_n_splits(self, *args: Any, **kargs: Any) -> int:
+    def get_n_splits(
+            self, *args: Any, **kargs: Any) -> int:
```
```suggestion
    def get_n_splits(self, *args: Any, **kargs: Any) -> int:
```

thibaultcordier commented on mapie/tests/test_regression.py:
```diff
@@ -361,6 +361,28 @@ def test_results_with_constant_sample_weights(strategy: str) -> None:
 
 
+@pytest.mark.parametrize("strategy", [*STRATEGIES])
+def test_results_with_constant_groups(strategy: str) -> None:
+    """
+    Test predictions when groups are None
+    or constant with different values.
+    """
```
The addition of test results with non-constant groups has its place alongside this test to also cover this use case.

pidefrem replied:
Sure. I'll add some more tests.

thibaultcordier commented on HISTORY.rst:
```diff
@@ -4,6 +4,9 @@ History
 
 ##### (##########)
 ------------------
+* Allow the use of `y` and `groups` arguments in cross validator methods `get_n_splits`
+  and `split` to enable more cv-split variants for :class:`MapieRegressor`
+  (e.g. :class:`GroupKFold`, stratified continuous split).
```
You can add that you also make this change in the classification file.

pidefrem replied:
Fixed by listing methods with signature change.

thibaultcordier reviewed (changes requested):
Wonderful! Thank you for this very welcome contribution. I don't have many comments to add but just a few to complete your proposal exhaustively:
- some style code suggestions
- As far as the tests are concerned, you are indeed testing whether MAPIE returns the same results using the same group (constant or None). But what happens if you use different groups like `np.concat([np.ones(shape=n_samples/2), 2*np.ones(shape=n_samples/2)])` (untested proposal).
- duplicate the tests in the classification test file because changes have been made in this part of the code.

pidefrem commented:
> Wonderful! Thank you for this very welcome contribution. I don't have many comments to add but just a few to complete your proposal exhaustively:
> 
> * some style code suggestions
> * As far as the tests are concerned, you are indeed testing whether MAPIE returns the same results using the same group (constant or None). But what happens if you use different groups like `np.concat([np.ones(shape=n_samples/2), 2*np.ones(shape=n_samples/2)])` (untested proposal).
> * duplicate the tests in the classification test file because changes have been made in this part of the code.

@thibaultcordier I fixed some issues and added some tests, tell me if it is ok now.

thibaultcordier commented:
> > Wonderful! Thank you for this very welcome contribution. I don't have many comments to add but just a few to complete your proposal exhaustively:
> > 
> > * some style code suggestions
> > * As far as the tests are concerned, you are indeed testing whether MAPIE returns the same results using the same group (constant or None). But what happens if you use different groups like `np.concat([np.ones(shape=n_samples/2), 2*np.ones(shape=n_samples/2)])` (untested proposal).
> > * duplicate the tests in the classification test file because changes have been made in this part of the code.
> 
> @thibaultcordier I fixed some issues and added some tests, tell me if it is ok now.

Hello @pidefrem, I'll check your changes this week. Thank you for contacting me about the review. I'll keep you informed.

thibaultcordier reviewed (approved):
Great! Thank you for this contribution, I don't have any feedback to give, ready to merge!

LacombeLouis commented on .gitignore:
```diff
@@ -77,6 +77,7 @@ doc/generated/
 target/
 
 # VSCode
+.venv
```
Please remove the environment from the MAPIE folder!
Thank you

pidefrem replied:
done

LacombeLouis reviewed (approved):
Hey @pidefrem,
Thank you for this PR! This will be extremely useful. One little modification would be to remove the environment from the MAPIE folder/GitHub. It can be saved outside the folder. 
If you can do this, we will be able to merge your PR asap!
Thank you again!

LacombeLouis reviewed (changes requested):
One little modification would be to remove the environment from the MAPIE folder/GitHub. It can be saved outside the folder.
